[ACP-77] Updates (#135)
avalanche-foundation/ACPs: ACPs/77-reinventing-subnets/README.md

By separating Subnet Validators from Primary Network Validators, a list of validator IPs to connect to (the functional bootstrappers of the Subnet) is no longer provided by simply connecting to the Primary Network Validators. However, the Primary Network can enable nodes tracking a Subnet to seamlessly connect to the Subnet Validators by tracking and gossiping Subnet Validator IPs. Subnets will not need to operate and maintain a set of bootstrappers and can continue to rely on the Primary Network for peer discovery.

### Setting a Subnet Manager

To create a Subnet, a `CreateSubnetTx` must be issued on the P-Chain. This transaction includes an `Owner` field which defines the key that must be used to authorize any validator set additions (`AddSubnetValidatorTx`) or removals (`RemoveSubnetValidatorTx`).

> To be a Permissionless Subnet, this `Owner` key must no longer have the ability to modify the Subnet's validator set. A `ConvertSubnetTx` must first be issued to explicitly convert a Subnet from Permissioned to Permissionless. This transaction will set the `(chainID, address)` pair that will manage the Subnet going forward. After `ConvertSubnetTx` is issued, the `Owner` from the `CreateSubnetTx` that created the Subnet will no longer have the ability to modify the Subnet's validator set.

Reviewer: Would a `[CreateChainTx`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/create_chain_tx.go#L36) need to be issued _before_ `ConvertSubnetTx`?

Author: > Would a CreateChainTx need to be issued before ConvertSubnetTx?

Yes

Author: > Also wondering if a Subnet can be permissionless from its creation or if CreateChainTx will eventually accept a Subnet Manager that is not Owner from the get go

Currently no. It starts as Permissioned then can be migrated to Permissionless. ACP-77 will probably stop here in terms of Subnet creation flow since it's already fairly involved. The UX streamlining you are describing would be better suited in a follow up ACP

> To provide maximal flexibility for Permissionless Subnets, the BLS multisignature approach in Avalanche Warp Messaging (AWM) is re-used to approve modifications to the Subnet's validator set. Using the `(chainID, address)` pair defined in the `ConvertSubnetTx`, a Warp Message with an [`AddressedCall`](https://github.com/ava-labs/avalanchego/tree/master/vms/platformvm/warp/payload#addressedcall) payload can be constructed.

Reviewer 2: just a clarification on wording is "BLS multisignature" the same as "BLS aggregate signature"? I haven't heard of it referred to as a multisig before.

Author: The proper term is multisignature: https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html

> To validate an `AddressedCall` payload in Avalanche Warp Messaging, the `(chainID, address)` pair is used to lookup the validators of `chainID` and verify that the BLS multi-sig includes a quorum (set to 67%) of the Subnet's validator set. The P-Chain will use Warp Messages with `AddressedCall` payloads to support modifications of the Subnet's validator set using this `(chainID, address)` pair (using the `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` defined later in the specification).

Reviewer 3: Should it be specified here that the `sourceAddress` of the `AddressedCall` payload is expected to have 0 length?

Author: The `sourceAddress` is not expected to have 0 length for `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx`. It is expected to have 0 length for the message from P-Chain -> Subnet which is addressed in the `Proof of Subnet Validator Set Change` section

> To validate an `AddressedCall` payload in Avalanche Warp Messaging, the `(chainID, address)` pair is used to lookup the validators of `chainID` and verify that the BLS multi-sig includes a quorum (set to 67%) of the Subnet's validator set. The P-Chain will use Warp Messages with `AddressedCall` payloads to support modifications of the Subnet's validator set using this `(chainID, address)` pair (using the `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` defined later in the specification).

Reviewer 4: does the chain referenced by chainID needs to belong to the subnet to be transformed? (it kind of seems so from the description `the (chainID,address) pair is used to lookup the validators of chainID`)

> To validate an `AddressedCall` payload in Avalanche Warp Messaging, the `(chainID, address)` pair is used to lookup the validators of `chainID` and verify that the BLS multi-sig includes a quorum (set to 67%) of the Subnet's validator set. The P-Chain will use Warp Messages with `AddressedCall` payloads to support modifications of the Subnet's validator set using this `(chainID, address)` pair (using the `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` defined later in the specification).

Reviewer 4: how is P-Chain using the chainID in the (chainID, address) pair when it receives a RegisterSubnetValidatorTx or a SetSubnetValidatorWeightTx, if it is the case that the chainID may not belong to the subnet to be transformed?

> To validate an `AddressedCall` payload in Avalanche Warp Messaging, the `(chainID, address)` pair is used to lookup the validators of `chainID` and verify that the BLS multi-sig includes a quorum (set to 67%) of the Subnet's validator set. The P-Chain will use Warp Messages with `AddressedCall` payloads to support modifications of the Subnet's validator set using this `(chainID, address)` pair (using the `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` defined later in the specification).

Author: The chain can be any chain. RegisterSubnetValidatorTx and SetSubnetValidatorWeightTx is only valid if the (chainID, address) is equivalent to what is stored on the P-Chain for the Subnet manager

> To validate an `AddressedCall` payload in Avalanche Warp Messaging, the `(chainID, address)` pair is used to lookup the validators of `chainID` and verify that the BLS multi-sig includes a quorum (set to 67%) of the Subnet's validator set. The P-Chain will use Warp Messages with `AddressedCall` payloads to support modifications of the Subnet's validator set using this `(chainID, address)` pair (using the `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` defined later in the specification).

Reviewer 4: Ok

#### ConvertSubnetTx

```golang
type ConvertSubnetTx struct {
    // Metadata, inputs and outputs
    BaseTx
    // ID of the Subnet to transform
    // Restrictions:
    // - Must not be the Primary Network ID
    Subnet ids.ID `json:"subnetID"`
    // Chain where the Subnet manager lives
    ChainID ids.ID `json:"chainID"`
    // Address of the Subnet manager
    Address []byte `json:"address"`
    // Authorizes this conversion
    SubnetAuth verify.Verifiable `json:"subnetAuthorization"`
}
```

Once this transaction is accepted, `AddSubnetValidatorTx` and `RemoveSubnetValidatorTx` are disabled on the Subnet. `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx` must be used to manage the Subnet's validator set going forward.

### Recovering a Subnet

> In the event that a Subnet has no validators, a valid BLS multi-signature cannot be produced. This situation can arise immediately after a Subnet is created or if all the Subnet's validators were removed from the validator set via `SetSubnetValidatorWeightTx` with `Weight = 0`.

Reviewer 3: ```suggestion
In the event that a Subnet has no validators, a valid BLS multi-signature cannot be produced. This is the case immediately after a Subnet is created, and can arise later if all the Subnet's validators are removed from the validator set.
```

Wording nit. On the second part, validators can be removed via `SetSubnetValidatorWeightTx` or `ExitValidatorSetTx` right? Don't think it's necessary to specify both here, but mostly don't think we should include one without the other.

> A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the `Owner` key defined in `CreateSubnetTx`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued.

Reviewer 5: Is `SubnetAuth` in `RecoverSubnetTx` and `ConvertSubnetTx` the same as `Owner`?

Reviewer: `SubnetAuth` is necessary for any transactions that require signing with an Owner Key. Many of the existing p-chain transactions already implement this, i.e. [CreateChainTx](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/create_chain_tx.go#L50). Here's an [example](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/wallet/chain/p/signer/visitor.go#L96) made with the p chain wallet.

Reviewer 5: Thanks for the context. I commented because it wasn't clear to me how `Owner` maps to the code. Any docs that describe the relationship between `Owner` and `SubnetAuth` would be helpful.

Reviewer: definitely need better docs on this. afaict, the [`verifySubnetAuthorization`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/executor/subnet_tx_verification.go#L53) uses [`getSubnetOwner`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/state/diff.go#L279) to get the owner from state. This was set by [`setSubnetOwner`](https://github.com/ava-labs/avalanchego/blob/a0f7e422bb8497186ebff434f12c4b957cec8d49/vms/platformvm/txs/executor/standard_tx_executor.go#L107) when `CreateSubnetTx` was executed.

> A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the `Owner` key defined in `CreateSubnetTx`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued.

Reviewer 6: Also shouldn't it be possible to change the Subnet Manager after the `ConvertSubnetTx`? e.g. to change the address of the manager?

> A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the `Owner` key defined in `CreateSubnetTx`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued.

Author: How does this sound to you?

```suggestion
A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the Subnet's current `Owner`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued. The only exception is for `Owner` rotation via the `TransferSubnetOwnershipTx`.
```

> A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the `Owner` key defined in `CreateSubnetTx`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued.

Author: > Also shouldn't it be possible to change the Subnet Manager after the ConvertSubnetTx? e.g. to change the address of the manager?

Rotating the manager address has a ton of issues that need to be worked out (e.g. how do you migrate the validator set and associated stake from the original address to the new address?). The original draft of ACP-77 had a manager rotation but it was removed because of that additional complexity it added to an already large ACP.

> A `RecoverSubnetTx` can be used in this situation to instantiate the Subnet's validator set using the `Owner` key defined in `CreateSubnetTx`. In all other situations, the `Owner` key is powerless after a `ConvertSubnetTx` is issued.

Reviewer 6: Sounds good!

It makes sense to remove the manager's address rotation. I agree with the added complexity.

> Each Subnet Validator will be instantiated with a Balance of `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` rounded down. For a `RecoverSubnetTx` to be valid, `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` must be >= the greater of 5 $AVAX or two weeks of the current fee. This prevents Subnet Validators from being added with too low of an initial balance where they become immediately delinquent based on the continous fee mechanism defined below. A Subnet Validator can leave at any time before the initial $AVAX is consumed and claim the remaining balance to the `ChangeOwner` defined in the transaction.

Reviewer 3: Super nit: Do we want to capitalize Validator? I don't think I've seen it capitalized elsewhere

Reviewer: should be lowercase

Author: I've been using uppercase here to follow "Primary Network Validator"

> Each Subnet Validator will be instantiated with a Balance of `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` rounded down. For a `RecoverSubnetTx` to be valid, `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` must be >= the greater of 5 $AVAX or two weeks of the current fee. This prevents Subnet Validators from being added with too low of an initial balance where they become immediately delinquent based on the continous fee mechanism defined below. A Subnet Validator can leave at any time before the initial $AVAX is consumed and claim the remaining balance to the `ChangeOwner` defined in the transaction.

Reviewer: ```suggestion
Each Subnet validator will be instantiated with a Balance of `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` rounded down. For a `RecoverSubnetTx` to be valid, `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)`(the initial Balance per validator) must be >= the greater of 5 $AVAX or two weeks of the current fee. This prevents Subnet validators from being added with too low of an initial balance where they become immediately delinquent based on the continuous fee mechanism defined below. A Subnet validator can leave at any time before the initial $AVAX is consumed and claim the remaining balance to the `ChangeOwner` defined in the transaction.
```

> Each Subnet Validator will be instantiated with a Balance of `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` rounded down. For a `RecoverSubnetTx` to be valid, `(sum($AVAX inputs) - sum($AVAX outputs) - TxFee) / len(Validators)` must be >= the greater of 5 $AVAX or two weeks of the current fee. This prevents Subnet Validators from being added with too low of an initial balance where they become immediately delinquent based on the continuous fee mechanism defined below. A Subnet Validator can leave at any time before the initial $AVAX is consumed and claim the remaining balance to the `ChangeOwner` defined in the transaction.

Reviewer: where is `Balance` input in `RecoverSubnetTx`?

Author: Good catch, added!

#### RecoverSubnetTx

```golang
type SubnetValidator struct {
    // Must be Ed25519 NodeID
    NodeID ids.NodeID `json:"nodeID"`
    // Weight of this validator used when sampling
    Weight uint64 `json:"weight"`
    // [Signer] is the BLS key for this validator.
    // Note: We do not enforce that the BLS key is unique across all validators.
    //       This means that validators can share a key if they so choose.
    //       However, a NodeID + Subnet does uniquely map to a BLS key
    Signer signer.Signer `json:"signer"`
    // Leftover $AVAX from the [Balance] will be issued to this
    // owner once it is removed from the validator set.

>     ChangeOwner fx.Owner `json:"changeOwner"`

Author: Any suggestions? I picked ChangeOwner to match existing nomenclature 🤷🏾‍♂️

>     ChangeOwner fx.Owner `json:"changeOwner"`

Reviewer 4: Yes surely. I can suggest something on the lines of ValidatorLeftOverBalanceOwner

}

> type RecoverSubnetTx struct {

Reviewer 4: maybe Recover is not the best name for the case of first using it, after a convert.  maybe the starting set of validators can be just set on Convert itself?

Author: Existing PoA Subnets should be able to keep their validator set through the ConvertSubnetTx so I didn't include the starting set on ConvertSubnetTx.

Recover signifies that this is a transaction that is only used as last-resort. Open to other suggestions

Reviewer 4: It seems the case that is going to be used for the typical flow, not as last resort. Maybe SetInitialSOV?

    BaseTx
    // Balance must be equal to the balance uint64 in the [Message].
    // Balance <= sum($AVAX inputs) - sum($AVAX outputs) - TxFee.
    Balance uint64 `json:"balance"`
    // [Signer] is the BLS key for this validator.
    // Note: We do not enforce that the BLS key is unique across all validators.
    //       This means that validators can share a key if they so choose.
    //       However, a NodeID does uniquely map to a BLS key
    Signer signer.Signer `json:"signer"`
    // Leftover $AVAX from the Subnet Validator's Balance will be issued to
    // this owner after it is removed from the validator set.
    ChangeOwner fx.Owner `json:"changeOwner"`
    // AddressedCall with Payload:
    //   - SubnetID
    //   - NodeID (must be Ed25519 NodeID)
    //   - Weight
    //   - Expiry
    //   - Ed25519 Signature over "[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]"
    Message warp.Message `json:"message"`
}
```

The `Message` field must be an `AddressedCall` with the payload:

```text
+-----------+----------+-----------+
|   codecID :   uint16 |   2 bytes |
+-----------+----------+-----------+
|    typeID :   uint32 |   4 bytes |

> +-----------+----------+-----------+

Reviewer 5: These two fields are not represented in the type spec comment above.

Reviewer: https://github.com/dhrubabasu/ACPs/pull/1#discussion_r1696100009

Reviewer: IMO it should be reflected in the type spec to avoid confusion, otherwise we will probably get more questions about this

Author: By virtue of being an `AddressedCall` payload, there is a `codecID` and `typeID` (Ref: https://github.com/ava-labs/avalanchego/tree/master/vms/platformvm/warp/payload#addressedcall)

Reviewer 3: The `codecID` and `typeID` for the `AddressedCall` payload of the Warp message are different than the `codecID` and `typeID` for the payload within the `AddressedCall` body though, no?

Author: Yes, that's correct. The `codecID` and `typeID` here are for the payload within the `AddressedCall`.

+-----------+----------+-----------+
                       | 150 bytes |
                       +-----------+
```

- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
- `typeID` is the payload type identifier and is `0x00000000` for this transaction
- `subnetID`, `nodeID`, and `weight` are for the Subnet Validator being added
- `expiry` is the time after which this message is invalid. After the P-Chain timestamp is past `expiry`, this Avalanche Warp Message can no longer be used to add the `nodeID` to the validator set of `subnetID`.

    `validationID` is defined as the SHA256 hash of the `Payload` of the `AddressedCall`. This SHA256 hash will be used for replay protection. Used `validationID`s will be stored on the P-Chain. If a `RegisterSubnetValidatorTx`'s `validationID` has already been used, the transaction will be considered invalid. To prevent storing an unbounded number of `validationID`s, the `expiry` is required to be no longer than 48 hours in the future of the time the transaction is issued on the P-Chain. Any `validationIDs` with `expiry` more than 48 hours in the past can be flushed from the P-Chain's state.

> - `signature` is the raw bytes of the Ed25519 signature over the concatenated bytes of `[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]`. This signature must correspond to the Ed25519 public key that is used for the `nodeID`. This approach prevents NodeIDs from being unwillingly added to Subnets. `balance` is the minimum initial $nAVAX balance that must be attached to the validator serialized as a uint64. A Subnet may choose to validate that the `signature` field is well-formed but it is not required.

Reviewer 3: ```suggestion
- `signature` is the raw bytes of the Ed25519 signature over the concatenated bytes of `[subnetID]+[nodeID]+[blsPublicKey]+[weight]+[balance]+[expiry]`. This signature must correspond to the Ed25519 public key that is used for the `nodeID`. This approach prevents NodeIDs from being unwillingly added to Subnets. `balance` is the minimum initial $nAVAX balance that must be attached to the validator serialized as a uint64. A Subnet may choose to validate that the `signature` field is well-formed, but it is not required.
```

Should we move this bullet point above the `validationID` description? Seems a little odd in the middle of the field descriptions.

Author: Removed `signature` from warp message to use `blsPublicKey` directly.

```text
+--------------+----------+-----------+
|      codecID :   uint16 |   2 bytes |
+--------------+----------+-----------+
|       typeID :   uint32 |   4 bytes |
+--------------+----------+-----------+
|     subnetID : [32]byte |  32 bytes |
+--------------+----------+-----------+
|       nodeID : [32]byte |  32 bytes |
+--------------+----------+-----------+
|       weight :   uint64 |   8 bytes |
+--------------+----------+-----------+
| blsPublicKey : [48]byte |  48 bytes |
+--------------+----------+-----------+
|       expiry :   uint64 |   8 bytes |
+--------------+----------+-----------+
                          | 134 bytes |
                          +-----------+
```

- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
- `typeID` is the payload type identifier and is `0x00000000` for this transaction
- `subnetID`, `nodeID`, `weight`, and `blsPublicKey` are for the Subnet Validator being added
- `expiry` is the time after which this message is invalid. After the P-Chain timestamp is past `expiry`, this Avalanche Warp Message can no longer be used to add the `nodeID` to the validator set of `subnetID`.

>     `validationID` is defined as the SHA256 hash of the `Payload` of the `AddressedCall`. This SHA256 hash will be used for replay protection. Used `validationID`s will be stored on the P-Chain. If a `RegisterSubnetValidatorTx`'s `validationID` has already been used, the transaction will be considered invalid. To prevent storing an unbounded number of `validationID`s, the `expiry` is required to be no longer than 48 hours in the future of the time the transaction is issued on the P-Chain. Any `validationIDs` with `expiry` more than 48 hours in the past can be flushed from the P-Chain's state.

Reviewer 7: Why do we wait 48 hours past the `expiry` to flush the `validationID`?

Author: for replay protection. If we flush it 24 hours after, somebody can reissue the same message corresponding to the `validationID` and it would be valid since `expiry` is 24 hours in the future.

Reviewer 7: I'm not really understanding. If we flush it 24 hours after the expiry, the expiry is 24 hours in the past, not in the future.

Author: Ah I understand the confusion- this was poorly worded on my part.
```suggestion
    `validationID` is defined as the SHA256 hash of the `Payload` of the `AddressedCall`. This SHA256 hash will be used for replay protection. Used `validationID`s will be stored on the P-Chain. If a `RegisterSubnetValidatorTx`'s `validationID` has already been used, the transaction will be considered invalid. To prevent storing an unbounded number of `validationID`s, the `expiry` is required to be no longer than 48 hours in the future of the time the transaction is issued on the P-Chain. Any `validationIDs` with an `expiry` in the past can be flushed from the P-Chain's state.
```

type SetSubnetValidatorWeightTx struct {
    // Metadata, inputs and outputs
    BaseTx
    // AddressedCall with Payload:
    //   - ValidationID (SHA256 of the AddressedCall Payload of the RegisterSubnetValidatorTx adding the validator)
    //   - Nonce
    //   - Weight
    Message warp.Message `json:"message"`
}
```

The `Message` field must be an `AddressedCall` with the payload:

```text
+--------------+----------+----------+
|      codecID :   uint16 |  2 bytes |
+--------------+----------+----------+
|       typeID :   uint32 |  4 bytes |
+-----------+----------+----------+
| validationID : [32]byte | 32 bytes |
+--------------+----------+----------+
|        nonce :   uint64 |  8 bytes |
+--------------+----------+----------+
|       weight :   uint64 |  8 bytes |
+--------------+----------+----------+
                          | 54 bytes |
                          +----------+
```

- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
- `typeID` is the payload type identifier and is `0x00000001` for this transaction

> - `validationID` is the SHA256 of the `Payload` of the `AddressedCall` in the `RegisterSubnetValidatorTx` adding the validator to the Subnet's validator set

Reviewer 5: Are the `codecID` and `typeID` included in the `SHA256`?

Author: YEs

- `weight` is the new `weight` of the validator

### Removing Subnet Validators

#### SetSubnetValidatorWeightTx (Weight = 0)

Issuing a `SetSubnetValidatorWeightTx` with `Weight` of `0` will remove the Subnet Validator from the Subnet's Validator set.

All state related to the Subnet Validator being removed will be removed from the P-Chain's active state (BLS key, NodeID etc). This validator can be newly added to the Subnet's validator set using the `RegisterSubnetValidatorTx` flow.

#### ExitValidatorSetTx

Subnet Validators can use `ExitValidatorSetTx` to exit the validator set without interacting with the Subnet. An Ed25519 Signature is required using the Subnet Validator's Ed25519 private key for this transaction to be valid. Remaining $AVAX in the Subnet Validator's `Balance` will be issued to the `ChangeOwner` defined when adding this validator to the validator set.

> This transaction can be issued on the P-Chain without explicit consent of the Subnet. It is expected that validators should be removed from the Subnet's validator set through a `SetSubnetValidatorWeightTx` with weight `0` by initiating Subnet Validator removal on the Subnet. However, the ability to exit a Subnet Validator set is critical for censorship-resistance and/or failed Subnets. If a validator ever wishes to stop participating in Subnet consensus, they will be able to do so through this transaction. This is enforced on the P-Chain to prevent Subnets from locking Validators into participating in consensus indefinitely. Note that this does not modify a Subnet's total staking weight, this transaction moves the Validator from Active to Inactive.

Reviewer 2: Wanted to ask for clarification on `Note that this does not modify a Subnet's total staking weight, this transaction moves the Validator from Active to Inactive.`

because above it says `Remaining $AVAX in the Subnet Validator's `Balance` will be issued to the `ChangeOwner`, so can't the left over avax be removed from the Subnet's total staking weight?

Reviewer 2: nvm I mixed up weight and balanace. IIUC, "total staking weight" is referring to the Subnet's defined staking calculations, and "balance" that is moved to ChangeOwner is referred to the P-Chain AVAX balance.

> This transaction can be issued on the P-Chain without explicit consent of the Subnet. It is expected that validators should be removed from the Subnet's validator set through a `SetSubnetValidatorWeightTx` with weight `0` by initiating Subnet Validator removal on the Subnet. However, the ability to exit a Subnet Validator set is critical for censorship-resistance and/or failed Subnets. If a validator ever wishes to stop participating in Subnet consensus, they will be able to do so through this transaction. This is enforced on the P-Chain to prevent Subnets from locking Validators into participating in consensus indefinitely. Note that this does not modify a Subnet's total staking weight, this transaction moves the Validator from Active to Inactive.

Reviewer 3: Nit: the capitalization of "Validator" is inconsistent here.

}
```

### Proof of Subnet Validator Set Change

To track whether a Subnet Validator addition/modification/removal occured on the P-Chain, Subnet Validators must be willing to sign an `AddressedCall` attesting to the validator set change. Since all Subnet Validators sync the P-Chain, only the Subnet Validators need to sign the `AddressedCall`, not all Primary Network Validators.

> Two `AddressedCall`s are defined with the below payloads. For each of them, the `sourceChainID` must be set to the P-Chain ID and the `sourceAddress` must be set to an empty byte array.

Reviewer 5: Consider clarifying that `sourceAddress` is in the `AddressedCall`, and `sourceChainID` is in the top-level Warp message.

The method of requesting is left unspecified as it can be more generic. A viable option for supporting this functionality is laid out in [ACP-118](../118-warp-signature-request/README.md) with the `SignatureRequest` message.

> #### SubnetValidatorRegistrationMessage

Reviewer 6: I am wondering if having 2 payloads is really needed.  
Can't we always use the weight update payload below? Even to check for successful registration with `nonce` = `0`.

Author: I don't think it's strictly needed. However, I think it's cleaner to have two separate ones to correspond to the two different warp messages used in `RegisterSubnetValidatorTx` and `SetSubnetValidatorWeightTx`.

```text
+--------------+----------+----------+
|      codecID :   uint16 |  2 bytes |
+--------------+----------+----------+
|       typeID :   uint32 |  4 bytes |
+--------------+----------+----------+
| validationID : [32]byte | 32 bytes |
+--------------+----------+----------+
|        valid :     bool |  1 byte  | 
+--------------+----------+----------+
                          | 39 bytes |
                          +----------+
```

- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
- `typeID` is the payload type identifier and is `0x00000002` for this message
- `validationID` is the SHA256 of the `Payload` of the `AddressedCall` in the `RegisterSubnetValidatorTx` adding the validator to the Subnet's validator set

> - `valid` indicates whether or not the `validationID` corresponds to a valid `AddressedCall` payload

Reviewer 6: IMO the name of the field is confusing as well. Maybe `registered` would make more sense?

> - `valid` indicates whether or not the `validationID` corresponds to a valid `AddressedCall` payload

Reviewer 3: Yeah it's definitely kinda odd because there are more than two possible states but we're using a `bool`.

The 3 states are:
1. Currently active validation period.
2. Not a currently active validation period, but could possibly become one in the future (not yet passed its registration expiry time)
3. Not a currently active validation period, and will never be an active validation period (due to the expire being passed)

No message should be willing to be signed for validation IDs in case 2, so there isn't a need represent it at all.

I agree `registered` is more clear though.

> #### SetSubnetValidatorWeightMessage

Reviewer 6: To follow the same phrasing as `SubnetValidatorRegistrationMessage`:
```suggestion
#### SubnetValidatorWeightUpdateMessage
```

- `codecID` is the codec version used to serialize the payload and is hardcoded to `0x0000`
- `typeID` is the payload type identifier and is `0x00000002` for this message
- `validationID` is the SHA256 of the `Payload` of the `AddressedCall` in the `RegisterSubnetValidatorTx` adding the validator to the Subnet's validator set

> - `registered` indicates whether or not the `validationID` corresponds to a valid `AddressedCall` payload. If true, `validationID` corresponds to an active validator. If false, `validationID` does not correspond to an active validator, and never will as the `expiry` in the `AddressedCall` payload is in the past.

Reviewer 6: What if the `validationID` is not active yet, but the `expiry` is not in the past?

Reviewer 3: In this case, the P-Chain nodes should refuse to the sign the message.

Author: added a note

By moving ownership of the Subnet's validator set from the P-Chain to the Subnet, Subnet creators have no restrictions on what requirements they have to join their Subnet as a validator. Any stake that is required to join the Subnet Validator set is locked on the Subnet. If a validator is removed from the Subnet Validator set via a `SetSubnetValidatorWeightTx` with weight `0` or an `ExitValidatorSetTx` on the P-Chain, the stake on the Subnet will continue to be locked. How each Subnet handles stake associated with the Subnet Validator is entirely left up to the Subnet and can be treated independently to what happens on the P-Chain.

This new relationship between the P-Chain and Subnets provides a dynamic where Subnets can use the P-Chain as an impartial judge to modify parameters (in addition to its existing role of helping to validate incoming Avalanche Warp Messages). If a Validator is misbehaving, the Subnet Validators can collectively generate a BLS multisig to reduce the voting weight of a misbehaving validator. This operation is fully secured by the Avalanche Primary Network (225M $AVAX or $8.325B at the time of writing).

> Follow-up ACPs could extend the P-Chain <> Subnet relationship to include parametrization of the 67% threshold to enable Subnets to choose a different threshold based on their security model (e.g. a simple majority of 51%).

Reviewer 2: just curious, but could the idea be to include the Warp quorum in either CreateSubnetTx or ConvertSubnetTx?

Author: Could be. Left it to a future ACP to explore this since the security considerations need to be weighed very carefully
